streamer/src/nonblocking/stream_throttle.rs

```diff
@@ -106,76 +131,54 @@ impl StakedStreamLoadEMA {
-            (max_load_in_ema_window * max_load_in_ema_window * u128::from(stake))
-                / (current_load * u128::from(total_stake));
-
-        let calculated_capacity = capacity_in_ema_window
-            * u128::from(STREAM_THROTTLING_INTERVAL_MS)
-            / u128::from(ema_window_ms);
-        let calculated_capacity = u64::try_from(calculated_capacity).unwrap_or_else(|_| {
-            error!(
-                "Failed to convert stream capacity {} to u64. Using minimum load capacity",
-                calculated_capacity
-            );
-            self.stats
-                .stream_load_capacity_overflow
-                .fetch_add(1, Ordering::Relaxed);
-            MIN_STREAMS_PER_THROTTLING_INTERVAL_FOR_STAKED_CONNECTION
-        });
-
-        cmp::max(
-            calculated_capacity,
-            MIN_STREAMS_PER_THROTTLING_INTERVAL_FOR_STAKED_CONNECTION,
-        )
-    }
-}
-
-pub(crate) fn max_streams_for_connection_in_throttling_duration(
-    peer_type: ConnectionPeerType,
-    total_stake: u64,
-    ema_load: Arc<StakedStreamLoadEMA>,
-) -> u64 {
-    match peer_type {
-        ConnectionPeerType::Unstaked => {
-            let max_num_connections =
-                u64::try_from(MAX_UNSTAKED_CONNECTIONS).unwrap_or_else(|_| {
+        match peer_type {
+            ConnectionPeerType::Unstaked => self.max_unstaked_load_in_throttling_window,
+            ConnectionPeerType::Staked(stake) => {
+                // If the current load is low, cap it to 25% of max_load.
+                let current_load = u128::from(cmp::max(
+                    self.current_load_ema.load(Ordering::Relaxed),
+                    self.max_staked_load_in_ema_window / 4,
+                ));
+
+                // Formula is (max_load ^ 2 / current_load) * (stake / total_stake)
+                let capacity_in_ema_window = (u128::from(self.max_staked_load_in_ema_window)
+                    * u128::from(self.max_staked_load_in_ema_window)
+                    * u128::from(stake))
+                    / (current_load * u128::from(total_stake));
+
+                let calculated_capacity = capacity_in_ema_window
+                    * u128::from(STREAM_THROTTLING_INTERVAL_MS)
+                    / u128::from(EMA_WINDOW_MS);
+                let calculated_capacity = u64::try_from(calculated_capacity).unwrap_or_else(|_| {
                     error!(
-                        "Failed to convert maximum number of unstaked connections {} to u64.",
-                        MAX_UNSTAKED_CONNECTIONS
+                        "Failed to convert stream capacity {} to u64. Using minimum load capacity",
+                        calculated_capacity
                     );
-                    500
+                    self.stats
+                        .stream_load_capacity_overflow
+                        .fetch_add(1, Ordering::Relaxed);
+                    self.max_unstaked_load_in_throttling_window
+                        .saturating_add(1)
```

Comment:
What is the reasoning for this "+1"?

Reply:
This is to ensure staked connection get atleast 1 more stream than the unstaked connections.
The previous value (8) was lower than the stream count for unstaked connections. Doing this computationally helps make sure that the number will be correct, in case we change the other parameters that control stream counts for unstaked connections.

Reply:
Can we add a little comments there?

Reply:
Good idea, will do